### PR TITLE
Add dependency graph

### PR DESCRIPTION
**Pull request by tanak**
Things to do:
  show tips
  reduce rendering of svg
  sometimes rect size is wrong
  resize rect size along dependencyDepth

Also fixed indent

----

Please open this URL to check diff.
https://github.com/CyberAgent/patriot-workflow-scheduler/compare/develop...tanak:add-dependency-graph?expand=1&w=1

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,12 +14,16 @@
...
     "grunt": "^0.4.5",
     "grunt-browserify": "^4.0.1",
     "grunt-este-watch": "^0.1.18",
     "react": "^0.14.6",
-    "react-router": "^2.0.0",
+    "react-dom": "^0.14.6",
     "react-modal": "^0.6.1",
-    "react-dom": "^0.14.6"
+    "react-router": "^2.0.0",
+    "react-tabs": "^0.5.3"
```
Comment (zenmyo) [inline package.json]: lodash seems to be missing.

File: src/main/jsx/job/jobView.jsx [hunk 2]
```diff
@@ -6,21 +7,271 @@ import ChangeJobStateForm from './common/changeJobStateForm';
...
   "COMMAND_CLASS", "update_id", "state", "priority", "start_datetime", "exec_host", "exec_node",
   "products", "requisites", "consumers", "producers"
   ];
 
+const tabIndex = {
+  "TAB_INDEX_DEFAULT": 0,
+  "TAB_INDEX_GRAPH": 1
+};
+
+var DependencyGraph = React.createClass({
```
Comment (zenmyo) [inline src/main/jsx/job/jobView.jsx]: IMO, the graph component should be moved to a separate file for readability.

File: src/main/jsx/job/dependencyGraph.jsx [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    if (Object.keys(job).length === 0) return;
+    
+    // set history to the original job
+    job.history = this.state.history;
+
+    var nodes = this.state.graphNodes;
+    nodes[job.job_id] = job;
+    this.setState({graphNodes: nodes});
+
+    this.setState({
```
Comment (zenmyo) [inline src/main/jsx/job/dependencyGraph.jsx]: Why is `setState` called twice?

File: src/main/jsx/job/jobView.jsx [hunk 4]
```diff
@@ -34,8 +53,18 @@ module.exports = React.createClass({
       }.bind(this));
     }.bind(this));
   },
+  handleTabSelect: function(index, last) {
+    this.setState({tabIndex: index});
+  },
+  handleChangeProducerGraphDepth: function(val) {
```
Comment (zenmyo) [inline src/main/jsx/job/jobView.jsx]: These two methods should be moved dependencyGraphs.

**Comment by zenmyo** [comment]
Some of `setState` invocations should be merged.
(e.g., `this.setState({graphNodes: nodes});` and `this.setState({graphEdges: edges});`)

File: lib/patriot/worker/servlet/job_api_servlet.rb [hunk 5]
```diff
@@ -43,6 +43,48 @@ class JobAPIServlet < Patriot::Worker::Servlet::APIServletBase
...
         end
 
+        # get dependency graph of the specified job
+        #
+        # @param [String] job_id
+        # @param [String] p_level a max number of producer dependency level to get
+        # @param [String] c_level a max number of consumer dependency level to get
+        get '/:job_id/graph' do
+          job_id = params[:job_id]
+          producer_depth = params['p_depth'].to_i || 2
```
Comment (zenmyo) [inline lib/patriot/worker/servlet/job_api_servlet.rb]: Since `nil.to_i = 0 ` , this does not work as expected.

**Comment by zenmyo** [comment]
I have tested it against real data and it seems to be working fine.
Thanks for the contribution.

**Comment by tanak** [comment]
:)

**Pull request merged**